src/sync/mutex.rs

```diff
@@ -94,15 +94,79 @@ impl<T: ?Sized> Mutex<T> {
                 mutex: self,
             })),
             Err(TryLockError::WouldBlock) => panic!("mutex state out of sync"),
-        }
+        };
+
+        // We need to let other threads in here so they may fail a `try_lock`
+        thread::switch();
+
+        result
     }
 
     /// Attempts to acquire this lock.
     ///
     /// If the lock could not be acquired at this time, then Err is returned. This function does not
     /// block.
     pub fn try_lock(&self) -> TryLockResult<MutexGuard<T>> {
-        unimplemented!()
+        let me = ExecutionState::me();
+
+        let mut state = self.state.borrow_mut();
+        trace!(holder=?state.holder, waiters=?state.waiters, "trying to acquire mutex {:p}", self);
+
+        if let Some(holder) = state.holder {
+            if holder == me {
+                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
+            }
+        }
+
+        // We don't need a context switch here. There are two cases to analyze.
+        // * Consider that we manage to acquire the lock, but that there is another thread that
+        //   also wants to acquire. At the last context switch (where we were chosen), the other
+        //   thread must have been already runnable and could have been chosen by the scheduler.
+        //   Then that other thread's acquire has a context switch that allows us to run into the
+        //   `WouldBlock` case.
```

Comment:
Does this same argument also work for the existing `thread::switch` on line 68 of the regular `lock` path? Basically, if you ever enter `lock`, there must have been an earlier context switch that chose to run you, and could instead have chosen to run someone else to win the race for the lock.

(I think the answer is no, because I just tried it, but I can't convince myself why)

Reply:
For regular `lock` we need to implement blocking. If the previous context switch decides who wins the race, then the other threads stay runnable and can be scheduled while the lock is still held. I think what that means is that we can do the context switch *only* in the case where we find the lock held by another thread.